src/types/type_param.rs

```diff
@@ -62,72 +114,103 @@ impl TypeArg {
         match self {
             TypeArg::Type(s) => Some(s.tag()),
             TypeArg::ClassicType(c) => Some(c.tag()),
+            TypeArg::HashableType(h) => Some(h.tag()),
             _ => None,
         }
     }
 }
 
-/// Checks a [TypeArg] is as expected for a [TypeParam]
-pub fn check_type_arg(arg: &TypeArg, param: &TypeParam) -> Result<(), TypeArgError> {
-    match (arg, param) {
-        (TypeArg::Type(_), TypeParam::Type) => Ok(()),
-        (TypeArg::ClassicType(_), TypeParam::ClassicType) => Ok(()),
-        (TypeArg::HashableType(_), TypeParam::HashableType) => Ok(()),
-        (TypeArg::List(items), TypeParam::List(ty)) => {
-            for item in items {
-                check_type_arg(item, ty.as_ref())?;
+/// A serialized representation of a value of a [CustomType]
+/// restricted to Hashable types.
```

Comment:
This would just *be* `CustomSerialized` were it not for enforcing this, which I think is an important sanity check (no `MyType<3.51f>` even by putting that in a CustomType, but `MyType<3>` or custom versions of are ok)